### PR TITLE
Introduce Semantics API

### PR DESCRIPTION
**Pull request by matklad**
Based on ideas in https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Fwg-rls-2.2E0/topic/Design.20question.3A.20identity.20in.20syntax.20trees

### CODE REVIEW

File: crates/ra_ide/src/goto_type_definition.rs [hunk 1]
```diff
@@ -33,6 +33,34 @@ pub(crate) fn goto_type_definition(
...
+    position: FilePosition,
+) -> Option<RangeInfo<Vec<NavigationTarget>>> {
+    let sema = hir::Semantics::new(db);
+
+    let file: ast::SourceFile = sema.file_syntax(position.file_id);
+    let token: SyntaxToken = pick_best(file.syntax().token_at_offset(position.offset))?;
+    let token: SyntaxToken = sema.descend_into_macros(token);
+
+    let (ty, node) = sema.ancestors_with_macros(token.parent()).find_map(|node| {
+        let ty = if let Some(expr) = ast::Expr::cast(node.clone()) {
```
Comment (kjeremy) [inline crates/ra_ide/src/goto_type_definition.rs]: `match_ast!`?

File: crates/ra_hir/src/semantics.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        sa.type_of_pat(self.db, &pat)
+    }
+
+    fn cache(&self, root_node: SyntaxNode, file_id: HirFileId) {
+        assert!(root_node.parent().is_none());
+        let mut cache = self.cache.borrow_mut();
+        let prev = cache.insert(root_node, file_id);
+        assert!(prev.is_none());
+    }
+    fn lookup(&self, root_node: &SyntaxNode) -> HirFileId {
```
Comment (kjeremy) [inline crates/ra_hir/src/semantics.rs]: I think this name should be more specific. Maybe `file_for` or something?

**Comment by kjeremy** [comment]
Seems like an improvement to me!

**Comment by matklad** [comment]
ported syntax highlighting to `Semantics`. It does look nicer!

File: crates/ra_hir/src/semantics.rs [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+use crate::{
+    db::HirDatabase, source_analyzer::ReferenceDescriptor, source_binder::ToDef, Function,
+    HirFileId, InFile, MacroDef, Module, Name, Origin, PathResolution, ScopeDef, SourceAnalyzer,
+    SourceBinder, StructField, Type, VariantDef,
+};
+use hir_def::TraitId;
+
+pub struct Semantics<'db, DB> {
+    pub db: &'db DB,
+    sb: RefCell<SourceBinder<'db, DB>>,
```
Comment (Veetaha) [inline crates/ra_hir/src/semantics.rs]: I wish we could avoid `RefCell` for `SourceBinder` here...

Comment (matklad) [reply crates/ra_hir/src/semantics.rs]: This is the perfect use-case for RefCell actually. Or, rather, it was an API mistake to not use `RefCell` in the `SourceBinder` in the first place. 

This type is not thread safe (as it stores syntax nodes) and in general is transient in nature. Additionally, what `RefCell` guards here are essentially caches. From the consumers pov, this looks like an immutable data structure. And `&T` APIs are much more convenient to use than `&mut T` APIs.

Comment (Veetaha) [reply crates/ra_hir/src/semantics.rs]: So should we use `RefCell<FxHashMap<...>>` here inside of `SourceBinder` instead?https://github.com/rust-analyzer/rust-analyzer/blob/1fe48a0115c24240f5a3e1b329e642f18e2715d6/crates/ra_hir/src/source_binder.rs#L23-L26

**Comment by matklad** [comment]
Yeah, both sa and sb should just go away as a result of this refactoring .

**Comment by matklad** [comment]
Another interesting API i've discovered during the work: `SemanticScope`, which encapsulates scoping info at the given point of code. resolving `ast::Path` works with `Semantics`, resolving `hir::Path` needs a `SemanticsScope`.

**Comment by kiljacken** [comment]
@matklad ´SemanticScope´ seems like it would play nicely with local imports, so :+1: for that

**Comment by flodiebold** [comment]
> Another interesting API i've discovered during the work: `SemanticScope`, which encapsulates scoping info at the given point of code. resolving `ast::Path` works with `Semantics`, resolving `hir::Path` needs a `SemanticsScope`.

Yeah, that sounds like a good solution :thinking:

File: crates/ra_assists/src/assist_ctx.rs [hunk 4]
```diff
@@ -74,29 +75,22 @@ pub(crate) type AssistHandler = fn(AssistCtx) -> Option<Assist>;
 /// Note, however, that we don't actually use such two-phase logic at the
 /// moment, because the LSP API is pretty awkward in this place, and it's much
 /// easier to just compute the edit eagerly :-)
-#[derive(Debug)]
+#[derive(Clone)]
 pub(crate) struct AssistCtx<'a> {
+    // FIXME: Rc does not make much sense here
+    pub(crate) sema: Rc<Semantics<'a, RootDatabase>>,
```
Comment (Veetaha) [inline crates/ra_assists/src/assist_ctx.rs]: It made sense when you put it here or what did you want to say?

**Comment by matklad** [comment]
I think this is ready for review @flodiebold !

The public API looks good to me: we now have a single `Semantics` type, which handles all the things. We also don't really use `HirFileId` in public API anymore, *except* for `HasSource` trait. 

The implementation is unfortunatelly still split between sa, sb and semantics :(

File: crates/ra_hir/src/semantics.rs [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+    resolver: Resolver,
+}
+
+impl<'a, DB: HirDatabase> SemanticsScope<'a, DB> {
+    pub fn module(&self) -> Option<Module> {
+        Some(Module { id: self.resolver.module()? })
+    }
+
+    /// Note: `FxHashSet<TraitId>` should be treated as an opaque type, passed into `Type
+    // FIXME: rename to visible_traits to not repeat scope?
```
Comment (flodiebold) [inline crates/ra_hir/src/semantics.rs]: seems a better name, yeah

File: crates/ra_syntax/src/algo.rs [hunk 6]
```diff
@@ -56,6 +56,11 @@ pub fn find_covering_element(root: &SyntaxNode, range: TextRange) -> SyntaxEleme
     root.covering_element(range)
 }
 
+pub fn least_common_ancestor(u: &SyntaxNode, v: &SyntaxNode) -> Option<SyntaxNode> {
```
Comment (edwin0cheng) [inline crates/ra_syntax/src/algo.rs]: Neat !

File: crates/ra_hir/src/semantics.rs [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+        let cache = self.cache.borrow();
+        cache.get(root_node).copied()
+    }
+
+    fn find_file(&self, node: SyntaxNode) -> InFile<SyntaxNode> {
+        let root_node = find_root(&node);
+        let file_id = self.lookup(&root_node).unwrap_or_else(|| {
+            panic!(
+                "\n\nFailed to lookup {:?} in this Semantics.\n\
+                 Make sure to use only query nodes, derived from this instance of Semantics.\n\
```
[the remark this answers is not in the document]
Comment (matklad) [reply crates/ra_hir/src/semantics.rs]: Yeah... A path forward is to add that `u32` to rowan, as we originally planned, and add an interned query for `TreeProvenance` or some such.

**Pull request merged**